ci: Build with --enable-werror by default, and document exceptions

hebasto opened the pull request:
This PR prevents introducing of new compiler warnings in the master branch, e.g., #19986, #20162.

fanquake commented on ci/test/00_setup_env_i686_centos.sh:
```diff
@@ -13,3 +13,7 @@ export DOCKER_PACKAGES="gcc-c++ glibc-devel.x86_64 libstdc++-devel.x86_64 glibc-
 export BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-reduce-exports --with-boost-process"
 export CONFIG_SHELL="/bin/dash"
+
+# gcc-c++-4.8.5 is buggy, e.g.,
```
I'd rather wait until the next branch off so that we don't have to worry about compilers we are going to drop support for anyways.

hebasto replied:
That's why this is a draft for now :)

fanquake reviewed (commented):
Please fill out the PR description. What's the motivation for this change.

practicalswift commented on configure.ac:
```diff
@@ -382,25 +382,10 @@ if test "x$enable_werror" = "xyes"; then
   if test "x$CXXFLAG_WERROR" = "x"; then
     AC_MSG_ERROR("enable-werror set but -Werror is not usable")
   fi
-  AX_CHECK_COMPILE_FLAG([-Werror=gnu],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=gnu"],,[[$CXXFLAG_WERROR]])
```
I see `-Werror` being removed here, but where is it re-added? :)

hebasto replied:
Here: https://github.com/bitcoin/bitcoin/blob/0179abea02576b4e4291d36d14a2915d3d3c477d/configure.ac#L385

practicalswift replied:
Heh, thanks! Literally on the same line :) Sorry.

practicalswift commented:
Concept ACK: deduplication is good

vasild reviewed (commented):
Approach ACK

MarcoFalke commented:
The build system changes should go separate from the test/ci changes. Are they required or orthogonal?

vasild commented:
[@hebasto, correct me if I am wrong]

@MarcoFalke, if they are going to go as separate PRs, then I think the CI change must go first and after that the `configure.ac` change. Otherwise we will get CI trying to compile with full `-Werror` and without `--enable-suppress-external-warnings` which will likely cause build failures.

hebasto commented:
> The build system changes should go separate from the test/ci changes. Are they required or orthogonal?

CI changes require build system changes.

hebasto commented on ci/test/06_script_a.sh:
```diff
@@ -21,7 +24,7 @@ DOCKER_EXEC mkdir -p "${BASE_BUILD_DIR}"
 
 BEGIN_FOLD configure
-DOCKER_EXEC "${BASE_ROOT_DIR}/configure" --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( (DOCKER_EXEC cat config.log) && false)
+DOCKER_EXEC "${BASE_ROOT_DIR}/configure" --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG $WERROR_CONFIG || ( (DOCKER_EXEC cat config.log) && false)
```
> Where does `WERROR_CONFIG` come from?

It's from an previous version of this pull. Going to remove. Thanks!

hebasto commented on configure.ac:
```diff
@@ -433,6 +420,10 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-register],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-register"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-implicit-fallthrough"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-copy],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-copy"],,[[$CXXFLAG_WERROR]])
+  if test -z "${host##*darwin*}"; then
+    dnl This is a temporary workaround. See https://github.com/bitcoin/bitcoin/pull/17919#issuecomment-674404780
+    AX_CHECK_COMPILE_FLAG([-Wunused-command-line-argument],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-unused-command-line-argument"],,[[$CXXFLAG_WERROR]])
+  fi
```
Yes, `-Wunused-command-line-argument` is clang specific. So the `if test...` is not required at all, right?

vasild replied:
Right!

vasild commented on configure.ac:
```diff
@@ -433,6 +420,8 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-register],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-register"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-implicit-fallthrough"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-copy],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-copy"],,[[$CXXFLAG_WERROR]])
+  dnl This is a temporary workaround. See https://github.com/bitcoin/bitcoin/pull/17919#issuecomment-674404780
+  AX_CHECK_COMPILE_FLAG([-Wunused-command-line-argument],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-unused-command-line-argument"],,[[$CXXFLAG_WERROR]])
```
When can this workaround be removed? It is not clear from the linked comment.

I think whenever `-Wno-foo` is added (silencing a class of warnings in Bitcoin Core code) that should link a PR that fixes those warnings or at least some concrete plan to ensure that the added `-Wno-foo` will not be forgotten and remain forever in `configure.ac`.

hebasto replied:
Right, but we have no a PR with a fix now ((

vasild reviewed (approved):
ACK 81ade9b33

vasild commented:
ACK 3a409f6

hebasto committed:
ci: Build with --enable-werror by default, and document exceptions

hebasto commented:
@fanquake @practicalswift @vasild @MarcoFalke 

Mind (re)reviewing this PR (at the early stage of 22.0 cycle)?

practicalswift commented:
cr ACK 2ca5156ab4309a0a1dd3e17e07a1f86a411589dd: patch looks correct!

MarcoFalke commented on src/qt/guiutil.cpp:
```diff
@@ -67,7 +68,7 @@ namespace GUIUtil {
 QString dateTimeStr(const QDateTime &date)
 {
-    return date.date().toString(Qt::SystemLocaleShortDate) + QString(" ") + date.toString("hh:mm");
+    return QLocale::system().toString(date.date(), QLocale::ShortFormat) + QString(" ") + date.toString("hh:mm");
```
this is not a `ci` change, but a gui refactor change? Should probably be separate

hebasto replied:
Done in https://github.com/bitcoin-core/gui/pull/137

hebasto replied:
^ merged :)

vasild commented:
The non-QT changes look good. We now use `Werror` also on CentOS after ditching gcc 4, good!

I don't feel confident enough to ACK the QT changes.

hebasto commented:
> I don't feel confident enough to ACK the QT changes.

Qt changes split into https://github.com/bitcoin-core/gui/pull/137 as [requested](https://github.com/bitcoin/bitcoin/pull/20182#discussion_r531428490) by @MarcoFalke.

vasild commented:
QT changes are still in this PR.

hebasto commented:
> QT changes are still in this PR.

Yes, as this PR is based on https://github.com/bitcoin-core/gui/pull/137 (OP has been updated accordingly).

hebasto commented:
Rebased 2ca5156ab4309a0a1dd3e17e07a1f86a411589dd -> a00dd1e1824f9cb38551e5be4bd1c7be10c9db16 ([pr20182.07](https://github.com/hebasto/bitcoin/commits/pr20182.07) -> [pr20182.08](https://github.com/hebasto/bitcoin/commits/pr20182.08)) on top of the merged https://github.com/bitcoin-core/gui/pull/137.

Now this PR contains non-Qt changes only.

MarcoFalke commented on ci/test/06_script_a.sh:
```diff
@@ -6,7 +6,10 @@
 export LC_ALL=C.UTF-8
 
-BITCOIN_CONFIG_ALL="--disable-dependency-tracking --prefix=$DEPENDS_DIR/$HOST --bindir=$BASE_OUTDIR/bin --libdir=$BASE_OUTDIR/lib"
+BITCOIN_CONFIG_ALL="--enable-suppress-external-warnings --disable-dependency-tracking --prefix=$DEPENDS_DIR/$HOST --bindir=$BASE_OUTDIR/bin --libdir=$BASE_OUTDIR/lib"
```
should the diff in this line (9) be part of the first commit?

hebasto replied:
Why do you think so? The first commit seems complete to me, and it shouldn't fire any error on CI, no?

MarcoFalke replied:
Ok, then I think they should be a separate pull request: https://github.com/bitcoin/bitcoin/pull/20182#issuecomment-713425454

Doing build system changes under the `ci: ` prefix seems confusing.

hebasto replied:
Done in #20544.

MarcoFalke replied:
20544 fails without `--enable-suppress-external-warnings`:

```
/tmp/cirrus-ci-build/depends/arm-linux-gnueabihf/include/boost/asio/impl/error.ipp:102:15: error:   by ‘virtual std::__cxx11::string boost::asio::error::detail::misc_category::message(int) const’ [-Werror=overloaded-virtual]
   std::string message(int value) const
               ^~~~~~~

hebasto replied:
> should the diff in this line (9) be part of the first commit?

You're right. Some jobs already have `--enable-werror` :man_facepalming:

hebasto commented:
Build system changes. i.e. the first commit, has been split out into a separated pr #20544.

vasild commented:
To resolve the dependency tangle with #20544 I think that:

* This PR should drop its first commit (the `configure.ac` changes, contained in #20544). However the changes to `ci/test/06_script_a.sh` which are now part of that first commit should be moved to the second commit (ci changes).
* This PR merged (it will not cause any ci breakage).
* #20544 should remove the changes to `ci/test/06_script_a.sh` which are contained in this PR and already merged.
* #20544 merged.

hebasto commented:
> To resolve the dependency tangle with #20544 I think that:
> 
>  * This PR should drop its first commit (the `configure.ac` changes, contained in #20544). However the changes to `ci/test/06_script_a.sh` which are now part of that first commit should be moved to the second commit (ci changes).
> 
>  * This PR merged (it will not cause any ci breakage).
> 
>  * #20544 should remove the changes to `ci/test/06_script_a.sh` which are contained in this PR and already merged.
> 
>  * #20544 merged (it will not cause any breakage because ci is now prepared by the already merged PR).

Something similar was in my mind, so done.

hebasto commented:
@MarcoFalke the assigned label could be switched from "Build system" to "Tests" now, right?

MarcoFalke commented on ci/test/00_setup_env_native_nowallet.sh:
```diff
@@ -12,3 +12,7 @@ export PACKAGES="python3-zmq clang-5.0 llvm-5.0"  # Use clang-5 to test C++17 co
+
+# clang < 6.0 emits the -Wmissing-braces warning when initializing an std::array with only one pair of braces:
+#  - https://bugs.llvm.org/show_bug.cgi?id=21629
+export NO_WERROR=1
```
I can't reproduce locally

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/20182#issuecomment-738023008).

maflcko reviewed (commented):
Not sure if this makes sense by itself in the current form. I failed to reproduce the compile failures locally.

hebasto commented:
Updated cbb234c0da2e98aa9fec06896a78d2ffb66ad6b7 -> a0489f3472f3799dc1ece32a59556fd239c4c14b ([pr20182.10](https://github.com/hebasto/bitcoin/commits/pr20182.10) -> [pr20182.11](https://github.com/hebasto/bitcoin/commits/pr20182.11)):
- removed unrelated changes
- rebased on top of the recent CI changes

hebasto commented on ci/test/00_setup_env_s390x.sh:
```diff
@@ -24,3 +24,4 @@ export TEST_RUNNER_ENV="LC_ALL=C"
 export RUN_FUNCTIONAL_TESTS=true
 export GOAL="install"
 export BITCOIN_CONFIG="--enable-reduce-exports --with-incompatible-bdb --with-boost-process"
+export NO_WERROR=1
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20182#issuecomment-738023008).

vasild reviewed (approved):
ACK 0e2d9ef5

I am fine with `NO_WERROR=1` in some of the `.sh` files as long as those files did not contain `--enable-werror` in their `BITCOIN_CONFIG` before this PR. I.e. they did not use werror before this PR and will not use it after this PR either.

Surely, the less `NO_WERROR=1` the better :)

MarcoFalke commented on ci/test/00_setup_env_win64.sh:
```diff
@@ -14,3 +14,4 @@ export RUN_FUNCTIONAL_TESTS=false
 export RUN_SECURITY_TESTS="true"
 export GOAL="deploy"
 export BITCOIN_CONFIG="--enable-reduce-exports --disable-gui-tests --without-boost-process"
+export NO_WERROR=1
```
I haven't tried this one yet. What is the error here?

hebasto replied:
https://cirrus-ci.com/task/6268532532969472
```
script/interpreter.cpp: In function ‘bool EvalChecksig(const valtype&, const valtype&, prevector<28, unsigned char>::const_iterator, prevector<28, unsigned char>::const_iterator, ScriptExecutionData&, unsigned int, const BaseSignatureChecker&, SigVersion, ScriptError*, bool&)’:
script/interpreter.cpp:429:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1plus: some warnings being treated as errors
```

MarcoFalke replied:
Thanks. Can confirm locally:

![Screenshot from 2020-12-03 18-29-59](https://user-images.githubusercontent.com/6399679/101065576-9a4bf300-3595-11eb-9a32-ea1d10c51c11.png)

MarcoFalke replied:
Could mention with a comment?

MarcoFalke commented:
ACK 0e2d9ef54cdf21fddcfec1d8188633fb159ea993 👈

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 0e2d9ef54cdf21fddcfec1d8188633fb159ea993 👈
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjnwQv/f0RCPeajlgbPOL9U/Wqu/v50H3e8UfBTrKadvu4rz560W2Lhco2duJTz
E1U2KcKridi8Q/1Xgg75Jz0qm6X9S6EOI+brvc4IWMFu2IrKyZa3rDcrUTGoA6cB
LIBiDpoQPkFaWJH1bAgQLDwF1VOHoQx01qXQkcZRSNAC1yT8yCVuJ7OWrUDDusw8
nY+X7MZpB01E70gfJhZWnJfBT1i3R21gk6Zv7iej0nEP81l3887V4IgtCqwCX+qS
iOQxcsz/4GHZNZ1fiU5mz5GpZBcikdn9jRsqnLkJWGzZr+9z3hCY1pUsKx4tooHu
8L0s8x1Q/OH3+SUHivJNbIZ/h8HCaABZXlFZqULbK6abH4Zx4YKSeGPLJYhmIxcv
H9hUUlBYlplWI3lPz0yNkGr4yC2aWCX6gpxJzLwhq+Nf3VKEXDtttp0tsUudvRBN
YfCMqk4e0wzh3Ce7iaZymSdZeT3AMfDl3TSvSKcgxo9CQ5mqW3Yzdq215BWoX2dV
xUa3Mvso
=wpNQ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bf23f2c7778536b97f325b2e84fde80d36846b764324b5b39cd6d0f4c94e45fa  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bf23f2c7778536b97f325b2e84fde80d36846b764324b5b39cd6d0f4c94e45faf0108a7aa846fc00c41aad9c5ab44424e95d08fff0103fc123bcae16efdd2acffde288d9cc0808f1045fc921a3f00874ee993a40a53fd10083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01041468ea13f7d622f64466ae42cb0695108f1045fc921a2f0087e671df04c303c390083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b338a6534ef0b40ec3d3d60e2a1ab93208f1045fc921a2f0080f8f8e9c5934a4e70083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010c37960fb2b40c07e7bf2285ac6f8b61608f1045fc921a2f008f8e22c4232ddbc6f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

hebasto commented:
Updated 0e2d9ef54cdf21fddcfec1d8188633fb159ea993 -> 2f6fe4e4e9e9e35e713c0a20cf891b023592110a ([pr20182.11](https://github.com/hebasto/bitcoin/commits/pr20182.11) -> [pr20182.12](https://github.com/hebasto/bitcoin/commits/pr20182.12), [diff](https://github.com/hebasto/bitcoin/compare/pr20182.11..pr20182.12)):
- addresses @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20182#discussion_r535441027):
> Could mention with a comment?

practicalswift commented:
cr ACK 2f6fe4e4e9e9e35e713c0a20cf891b023592110a: patch looks correct

vasild reviewed (approved):
ACK 2f6fe4e

MarcoFalke commented:
re-ACK 2f6fe4e4e9 🏏

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 2f6fe4e4e9 🏏
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhqegv7Bn3eV9QYxtVtRWO0JksQf+6awouqP9K4uHCR4Wkq/xG1yFw6+fMlm9yw
S/Zzv8tTdBfu3uwsqLGamAEd7lMrsN+nvqcnrl0gTOyaoOpKFfAZTOTD0kgF8zKI
n8323jIjYkOdCgLGsaDRgaHcQRZUoa9WRSr6SyavhXtBx1vKjeAGbWpVHt/knpG4
hNhqDO3gTxF6S6Bt7nSj7WkG9CTj1ZzY1yJJe6tFE5xIfefRC5LHVBhwfFaRLa4t
P4K9TPJnlbB7fJyf2/o80FKWNR4nLHsL/W218PQ58Iq8tEtvh2NcIwEcBHq0/sSV
Nn/J8S6eghuR2bKwKS8XpkR/4Y0MfFBRHeFWDKhGCk3QY0KB52lFwWOpDcOPYnvN
Gr79QqZJZs3Hghl5pha15yF+C8LPzWECs7Od7FAZJct+8Pf0J/M8lTDdZ4Y2VOpR
kdQG+v5tdLAiMdqNndtBYvJ7aKJaKaYFGISz8r+JrVYlMaBfD4/GZwETzxGPm4tX
GygUgrQE
=gz7a
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f9a4b582f6420d97a683d6112df2a75c1b6bc10c52fbb86286062edb48eef81d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f9a4b582f6420d97a683d6112df2a75c1b6bc10c52fbb86286062edb48eef81df0103910ec5d1ab27e60ff010fa396e79b7f08fff01006215a381ddad8bafb4dcb794a70219008f0208e74a22132fa546c6be4d6bbdea636254a51aad28e020c4a133e204c89679e3908f1045fc9d5e0f008b1c83e7637ac5ad90083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010139c88c96d364704e0fa99826924e3bd08f020e47b9c459e4a8feed29a95f84e41f77aea65bfc127e21601cb5d752aa044126008f0201f5948080942c722b5e5d06b7f61be552dfa01ba36653bed7a6bb548967c06ea08f0201ef8281330c83ea91a2ac63e4e0586c6829c6bbe08cf43febfff1b6a0c99c75708f1045fc9d5e0f0083f03de0dd8d8b5a00083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0104c9e1e829e8b2c23bdbdde7e0006b65f08f0207781f55bb5e019aded6d07b38645e4f1ce7438a40b0a0dce5be478f61cd9e07108f120a85efa9c935d1ff16f3ae0002b769f8eba72d614ba2e217da1905a76b8bcc7f508f1045fc9d5e0f0088bb7e4da7d30f90c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010844ddb437949b56e24f82c506d6ee89408f1045fc9d5e0f0086d7a51bdd777d7850083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
